Add PMP::longest_edge()

hoskillua opened the pull request:
## Summary of Changes
I have implemented the `longest_edge()` function requested in this [issue](https://github.com/CGAL/cgal/issues/5799). I have tested it on simple input and it works fine.

## Release Management

* Affected package(s): PMP
* Issue(s) solved (if any): implemented #5799

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -211,6 +211,54 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+  *
+  * Finds the longest edge of a given polygon mesh.
+  *
+  * @tparam PolygonMesh a model of `HalfedgeGraph`
```
```suggestion
  * @tparam PolygonMesh a model of `HalfedgeListGraph`
```

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -211,6 +211,54 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+
+    auto edge_range = edges(pmesh);
+
+    CGAL_assertion(edge_range.size() > 0);
```
`boost::graph_traits<PolygonMesh>::null_edge()` could be returned if the mesh is empty (testing `edge_range.begin()==edge_range.end()`)

hoskillua replied:
@sloriot  Thank you for the feedback. I am a bit confused. I assume you mean that I should replace the assertion with a condition `(edge_range.begin()==edge_range.end())` and return `boost::graph_traits<PolygonMesh>::null_edge()`. Is that right?
I also added a check on the returned variable from the `max_element()` before dereferencing it for safety.  I added these updates and your suggestions to new commit.

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -211,6 +211,54 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+/**
+  * \ingroup PMP_measure_grp
+  *
+  * Finds the longest edge of a given polygon mesh.
```
```suggestion
  * returns the longest edge of a given polygon mesh.
```

sloriot commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -211,6 +211,54 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+  * @tparam PolygonMesh a model of `HalfedgeGraph`
+  * @tparam NamedParameters a sequence of \ref bgl_namedparameters "Named Parameters"
+  *
+  * @param pmesh the polygon mesh to which `h` belongs
```
there is no `h` mentioned above. Copy-paste issue?

hoskillua replied:
Indeed. Thank you for the correction.

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -211,6 +211,63 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+        return boost::graph_traits<PolygonMesh>::null_edge();
+
+    auto edge_reference = std::max_element(edge_range.begin(), edge_range.end(), [&pmesh, &np](auto l, auto r) {
+        return squared_edge_length(l, pmesh, np)
```
This should use [CompareDistance_3](https://doc.cgal.org/latest/Kernel_23/classKernel_1_1CompareDistance__3.html#a2ee64f4b6e1ac24eab6935efcf9d22ca)

afabri commented:
What is missing:   
A line in the Reference Manual [index page](https://github.com/CGAL/cgal/blob/master/Polygon_mesh_processing/doc/Polygon_mesh_processing/PackageDescription.txt#L222)
A line in the [change log](https://github.com/CGAL/cgal/blob/master/Installation/CHANGES.md)

hoskillua commented:
@afabri, I updated the Reference Manual and change log files. about testing, I added calls to the function in [measure_test.cpp](https://github.com/CGAL/cgal/pull/6496/files#diff-63e7d31cdc21d74d4d80bb85cece9a712a52f2f05afb4a6732299eaeaae6d486) and it ran as expected on these tests. But I can also make a testing program for this function if required.

sloriot commented:
There is a test failing [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.5-Ic-87/Polygon_mesh_processing/TestReport_monet_x86_64-apple_bigsur_clang1200-release-64bits.gz). Basically it says that the edge iterator of Polyhedron is not detected as a `ForwardIterator`. It might be a bug on our side with the definition of the edge iterator type.

janetournois commented:
Hello, what is the status of this PR?
I can't find the associated small feature.

sloriot commented:
@janetournois from what I remember there is a technical issue that I wasn't able to easily workaround at the time I looked at it. Let me have a fresh look.

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -217,6 +217,75 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+
+    Geom_traits gt = choose_parameter<Geom_traits>(get_parameter(np, internal_np::geom_traits));
+
+    auto edge_range = edges(pmesh);
```
This should be `const&`  as a reference to an array is potentially returned.
Or we should  test   `edges(pmesh).empty()`

afabri commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:
```diff
@@ -217,6 +217,75 @@ squared_edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e
+        return res == SMALLER;
+    });
+
+    // if edge_reference is not derefrenceble
```
I  think that cannot happen as the range is not empty.

sloriot commented:
Successfully tested in CGAL-6.1-Ic-45

sloriot commented:
@MaelRL @afabri considering that https://github.com/CGAL/cgal/pull/8676 add minmax_dihedral_angle, shouldn't we change this PR to add `minmax_edge_length()`? (`minmax` comes from `std::minmax_element`)
